File: leetcode/54.螺旋矩阵/54.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> res = new ArrayList<Integer>();
        int left = 0, right = matrix[0].length - 1, top  = 0, buttom = matrix.length -1;
        while(left <= right && top <= buttom){
            for(int i = left; i <= right; i++){
                res.add(matrix[top][i]);
            }
            top++;
            for(int i = top; i <= buttom; i++){
                res.add(matrix[i][right]);
            }
            right--;
            if(left <= right && top <= buttom){
                for(int i = right; i>= left; i--){
                    res.add(matrix[buttom][i]);
                }
                buttom--;
                for(int i = buttom; i>= top; i--){
                    res.add(matrix[i][left]);
                }
                left++;
            }

        }
        return res;
    }
}

File: leetcode/54.螺旋矩阵/54.py
class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        res = list()
        left, right, top, buttom = 0, len(matrix[0]) - 1, 0, len(matrix) - 1
        while left <= right and top <= buttom:
            for i in range(left, right + 1):
                res.append(matrix[top][i])
            top += 1
            for i in range(top, buttom + 1):
                res.append(matrix[i][right])
            right -= 1
            if left <= right and top <= buttom:
                for i in range(right, left - 1, -1):
                    res.append(matrix[buttom][i])
                buttom -= 1
                for i in range(buttom, top - 1, -1):
                    res.append(matrix[i][left])
                left += 1
        return res
        